feature : 이메일 기반 회원가입, 로그인 기능 구현

minjee2758 opened the pull request:
## 작업 내용
- 이메일 기반 회원가입 기능 구현
- 이메일 기반 로그인 기능 구현

---

## 변경 사항
### 1. AuthController
(1) `signup` 메서드 추가 (회원가입)
- 입력 받아야하는 값 :  이메일, 비밀번호, 비밀번호확인, 이름, 별명, 나이


(2) `signin` 메서드 추가 (로그인)
- 입력 받아야하는 값 : 이메일, 비밀번호
- 구현한 주요 로직, 클래스, 메서드 등을 bullet 형식으로 기술해주세요.

### 2. UserRepositoryImpl
- `createUser(User user)` 메서드 추가
 : DB에 유저 저장 (JPA에서 save()와 동일)
- `findByEmail(String email)` 메서드 추가
: 이메일로 DB에서 유저 찾기

### 3. UserService
- `signup()` 회원가입 
   : 도메인 서비스에서 유저 존재여부, 비밀번호 인코딩, DB에 유저 저장 로직 수행

- `signin` 로그인
   : 도메인 서비스에서 이메일로 유저 찾고 비밀번호 대조 로직 수행
-> 토큰 발급

---

## 트러블 슈팅
    - 문제: dto 매핑 실패
    - 해결:  import된 `@RequestBody`가 Swagger꺼였음
    - import org.springframework.web.bind.annotation.RequestBody;로 변경

- 메서드 위치에 대해 고밍을 많이 했던 것 같습니다 ..ㅎ..ㅎ


---

## 해결해야 할 문제

- 메서드나 클래스의 위치가 이상하거나 주입 위치가 잘못된 곳이 있다면 말씀해주세요!!
- 회원가입 시 꼭 토큰 발급을 해야하는지 고민중입니다

---

## 참고 사항

- 회원가입때 유저의 모든 정보를 입력받지 않고, 가입 후 나중에 내 정보 수정에서 블로그링크와 같은 부가적 정보를 받을 예정

---

### 코드 리뷰 전 확인 체크리스트

- [ ]  불필요한 콘솔 로그, 주석 제거
- [ ]  커밋 메시지 컨벤션 준수 (`type : `)
- [ ]  기능 정상 동작 확인


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **신규 기능**
  - 회원가입 및 로그인 API가 추가되어 사용자가 이메일, 비밀번호 등으로 회원가입 및 로그인을 할 수 있습니다.
  - JWT 기반 인증 및 인가가 적용되어, 토큰을 통한 사용자 인증과 역할별 접근 제어가 가능합니다.
  - 비밀번호는 보안 강화를 위해 bcrypt로 암호화됩니다.
  - 회원가입 및 로그인 요청/응답에 대한 데이터 유효성 검증이 강화되었습니다.
  - 인증된 사용자 정보를 컨트롤러에서 쉽게 주입받을 수 있는 기능이 추가되었습니다.
  - JWT 필터가 도입되어 요청별 토큰 검증 및 권한 확인이 이루어집니다.
  - 사용자 도메인 서비스 및 저장소가 구현되어 사용자 관리가 체계화되었습니다.
  - 커스텀 예외 및 인증 관련 어노테이션이 추가되어 인증 흐름이 개선되었습니다.

- **버그 수정**
  - 없음

- **기타**
  - 사용자 엔티티 및 관련 서비스, 예외 처리 구조가 개선되었습니다.
  - 프로젝트 의존성에 bcrypt, JWT 라이브러리 및 macOS ARM64 네이티브 DNS 리졸버가 추가되었습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

minjee2758 committed:
feature : jwt 기본 세팅

minjee2758 committed:
feature : jwt 기본 세팅

minjee2758 committed:
feature : 회원가입, 로그인 기능 구현

minjee2758 committed:
feature : 회원가입, 로그인 기능 구현

minjee2758 committed:
유저 로그인 및 토큰 발급 기능 개발

minjee2758 committed:
유저 로그인 및 토큰 발급 기능 개발

minjee2758 committed:
refactor : 불필요한 import 삭제

minjee2758 committed:
refactor : 불필요한 import 삭제

minjee2758 committed:
docs : 서비스 주석 추가

minjee2758 committed:
Merge branch 'feature/authJwt' of https://github.com/ezcode-my/backend into feature/authJwt

chat26666 commented on src/main/java/org/ezcode/codetest/application/usermanagement/auth/dto/signin/SigninRequest.java:
```diff
@@ -0,0 +1,24 @@
+public class SigninRequest {
+	@Email(message = "이메일은 공백일 수 없습니다")
+	@NotBlank
+	private String email;
```
나중에 입력값 크기 검증 로직이 들어가면 좋을듯합니다

minjee2758 replied:
바로 반영하겠습니다!

chat26666 commented on src/main/java/org/ezcode/codetest/application/usermanagement/auth/service/AuthService.java:
```diff
@@ -0,0 +1,80 @@
+public class AuthService {
+
+	private final UserDomainService userDomainService;
+	private final JwtUtil jwtUtil;
```
JwtUtil 은 infra layer에 만드신 실제 jwt util 구현체 클래스 인가요?  JwtUtil 을 application service 에서 사용하실 때는 Port 패키지에 JwtUtil 에 대한 추상화 인터페이스를 만드시고 구현체가 그걸 implements 해서 사용하게끔 변경하시면 될듯합니다.

minjee2758 replied:
확인했습니다! application의 usermannagement 패키지 하위의 port에 JwtUtil를 인터페이스 만들고, 이를 infra에서 implements해 구현할 수 있도록 수정했습니다(JwtUtilImpl)!!

minjee2758 commented on src/main/java/org/ezcode/codetest/domain/user/model/entity/User.java:
```diff
@@ -66,24 +66,37 @@ public class User extends BaseEntity {
+
 	public void setDeleted(boolean deleted) {
 		this.isDeleted = true;
 	}
```
우리는 삭제 상태를 boolean으로 하기로 해가지고 괜찮다 바보토끼야~~ 
-> 변수 안받고 상태 변경할 수 있게 리팩토링
```java
	public void setDeleted() {
		this.isDeleted = true;
	}
```

minjee2758 commented on src/main/java/org/ezcode/codetest/domain/user/repository/UserRepository.java:
```diff
@@ -0,0 +1,13 @@
+import org.ezcode.codetest.domain.user.model.entity.User;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
ㅇ ㅓ ㅋ ㅣ

minjee2758 commented on src/main/java/org/ezcode/codetest/common/config/FilterConfig.java:
```diff
@@ -0,0 +1,26 @@
+	private final JwtUtil jwtUtil;
+
+	@Bean
+	public FilterRegistrationBean<JwtFilter> jwtFilter(){
+		FilterRegistrationBean<JwtFilter> filterRegistrationBean = new FilterRegistrationBean<>();
+		filterRegistrationBean.setFilter(new JwtFilter(jwtUtil));
+		filterRegistrationBean.addUrlPatterns("/*");
+
+		return filterRegistrationBean;
+
```
이거는 나중에 개발 다 되면 적용 예정

minjee2758 commented on src/main/java/org/ezcode/codetest/infrastructure/security/jwt/JwtUtil.java:
```diff
@@ -1,7 +1,72 @@
+		key = Keys.hmacShaKeyFor(secretKey.getBytes());
+	}
+
+	public String createToken(Long userId, String email, UserRole userRole, String username, String nickname) {
+		Date date = new Date();
+		log.info("토큰 생성 시작");
+
+		return BEARER_PREFIX +
+			Jwts.builder()
+				.setSubject(String.valueOf(userId))
+				.claim("email", email)
+				.claim("username", username)
+				.claim("nickName", nickname)
+				.claim("userRole", userRole)
+				.setExpiration(new Date(date.getTime() + TOKEN_EXPIRATION_TIME))
+				.setIssuedAt(date)
+				.signWith(key, signatureAlgorithm)
+				.compact();
+	}
```
오

minjee2758 committed:
feature : password와 passwordConfirm을 비교하여 다르면 에러를 반환하는 예외 추가

minjee2758 committed:
refactor : application에 Jwt 인터페이스 구현

application의 AuthService에 Jwt관련 로직을 주입할 때, infra에서 바로 주입하지 말고, application에 JwtUtil interface를 만들어 이를 주입받을 수 있도록 함. JwtUtil 인터페이스를 infra의 JwtUtilImpl에 implements하여 구현

minjee2758 commented on src/main/java/org/ezcode/codetest/domain/user/service/UserDomainService.java:
```diff
@@ -1,8 +1,44 @@
+			.orElseThrow(() -> new AuthException("사용자를 찾을 수 없습니다."));
+	}
+
+	public void userPasswordCheck(String email, String password) {
+		User user = userRepository.findByEmail(email)
+			.orElseThrow(() -> new AuthException("사용자를 찾을 수 없습니다."));;
+
+		if (passwordEncoder.matches(password, user.getPassword())) {
+			throw new AuthException("비밀번호가 일치하지 않습니다");
+		}
+	}
```
오잉 잘되는데...?

minjee2758 committed:
refactor : 메서드 명 변경

유저 존재 existUser 메서드이므로 isPresent()로 검증하는게 맞음

minjee2758 committed:
refactor : 토큰 만료 시간 수정 (밀리초)

minjee2758 committed:
feature : Mac 환경에서 빌드시 필요한 의존성 추가

minjee2758 committed:
feature : Jwt예외 추가

토큰 만들때 : 토큰에 필요한 매개변수가 하나라도 없으면 예외. / 토큰 추출할 때 : 유효한 토큰이 아니면 예외

minjee2758 committed:
refactor : 로그 수정
